File: ros_ws/src/xela_description/xela_description/finger.rb
require 'erb'
require 'matrix'
require_relative 'list_to_string'


class LinkP4
  attr_reader :prefix

  def initialize(prefix)
    @prefix = prefix
  end

  # For now this template generates a single link (p4_unified),
  # matching `xela_description/robot.urdf`.
  def link_name
    "#{prefix}_p4_unified"
  end

  def inertial
    {
      origin: { xyz: [0.0342889, -0.0265621, -0.0145004], rpy: [0.0, 0.0, 0.0] },
      mass: 0.044,
      inertia: {
        ixx: 6.90801e-06,
        ixy: 2.26365e-06,
        ixz: 1.79475e-07,
        iyy: 1.04766e-05,
        iyz: -1.82708e-07,
        izz: 1.10431e-05
      }
    }
  end

  def visual
    {
      origin: { xyz: [0.0048, -0.0, -0.0145], rpy: [1.5708, 1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p4_unified.stl' } },
      material: {
        name: 'p4_unified_material',
        color: { rgba: [0.972549, 0.529412, 0.00392157, 1.0] }
      }
    }
  end

  def collision
    {
      origin: { xyz: [0.0048, -0.0, -0.0145], rpy: [1.5708, 1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p4_unified.stl' } }
    }
  end
end


class LinkP3
  attr_reader :prefix

  def initialize(prefix)
    @prefix = prefix
  end

  def link_name
    "#{prefix}_p3_unified"
  end

  def inertial
    {
      origin: { xyz: [-0.000258994, 0.00848001, 0.0133488], rpy: [0.0, 0.0, 0.0] },
      mass: 0.032,
      inertia: {
        ixx: 3.54425e-06,
        ixy: -1.6457e-08,
        ixz: -1.5221e-08,
        iyy: 2.67248e-06,
        iyz: -1.71388e-07,
        izz: 3.5384e-06
      }
    }
  end

  def visual
    {
      origin: { xyz: [0.0, 0.0075, 0.0], rpy: [-1.5708, -1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p3_unified.stl' } },
      material: {
        name: 'p3_unified_material',
        color: { rgba: [0.972549, 0.529412, 0.00392157, 1.0] }
      }
    }
  end

  def collision
    {
      origin: { xyz: [0.0, 0.0075, 0.0], rpy: [-1.5708, -1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p3_unified.stl' } }
    }
  end
end

class LinkP2
  attr_reader :prefix

  def initialize(prefix)
    @prefix = prefix
  end

  def link_name
    "#{prefix}_p2_unified"
  end

  def inertial
    {
      origin: { xyz: [0.0304899, -0.00688132, -0.0146269], rpy: [0.0, 0.0, 0.0] },
      mass: 0.037,
      inertia: {
        ixx: 8.11059e-06,
        ixy: 5.66349e-07,
        ixz: 1.88699e-07,
        iyy: 6.98155e-06,
        iyz: -9.6344e-08,
        izz: 9.0141e-06
      }
    }
  end

  def visual
    {
      origin: { xyz: [0.00905, -0.0, -0.0145], rpy: [1.5708, 1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p2_unified.stl' } },
      material: {
        name: 'p2_unified_material',
        color: { rgba: [0.768627, 0.886275, 0.952941, 1.0] }
      }
    }
  end

  def collision
    {
      origin: { xyz: [0.00905, -0.0, -0.0145], rpy: [1.5708, 1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/p2_unified.stl' } }
    }
  end
end

class LinkFingertop
  attr_reader :prefix

  def initialize(prefix)
    @prefix = prefix
  end

  def link_name
    "#{prefix}_fingertop_unfied"
  end

  def inertial
    {
      origin: { xyz: [-0.00193531, -0.0359351, 0.0145854], rpy: [0.0, 0.0, 0.0] },
      mass: 0.016,
      inertia: {
        ixx: 1.0779e-05,
        ixy: 1.57487e-07,
        ixz: -2.8587e-08,
        iyy: 5.55855e-06,
        iyz: -1.64744e-07,
        izz: 1.00656e-05
      }
    }
  end

  def visual
    {
      origin: { xyz: [-0.00135925, -0.025, 0.0145], rpy: [-1.5708, -1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/fingertop_unfied.stl' } },
      material: {
        name: 'fingertop_unfied_material',
        color: { rgba: [0.917647, 0.917647, 0.917647, 1.0] }
      }
    }
  end

  def collision
    {
      origin: { xyz: [-0.00135925, -0.025, 0.0145], rpy: [-1.5708, -1.5708, 0.0] },
      geometry: { mesh: { filename: 'package://assets/fingertop_unfied.stl' } }
    }
  end
end

class JointP4P3
  attr_reader :prefix
  attr_reader :offset

  def initialize(prefix, offset)
    @prefix = prefix
    @offset = offset
  end

  def joint_name
    "#{prefix}_rot"
  end

  def joint_type
    'revolute'
  end

  def origin
    { xyz: [0.0326 + offset, -0.0123, -0.0145], rpy: [-1.5708, -1.5708, 0.0] }
  end

  def parent_link_name
    "#{prefix}_p4_unified"
  end

  def child_link_name
    "#{prefix}_p3_unified"
  end

  def axis
    [0.0, 0.0, 1.0]
  end

  def limit
    { effort: 10, velocity: 10, lower: -1.0472, upper: 1.0472 }
  end
end


class JointP3P2
  attr_reader :prefix
  attr_reader :offset

  def initialize(prefix, offset)
    @prefix = prefix
    @offset = offset
  end

  def joint_name
    "#{prefix}_pip"
  end

  def joint_type
    'revolute'
  end

  def origin
    { xyz: [0.0145 + offset, 0.015, 0.013], rpy: [1.5708, 0.0, 1.5708] }
  end

  def parent_link_name
    "#{prefix}_p3_unified"
  end

  def child_link_name
    "#{prefix}_p2_unified"
  end

  def axis
    [0.0, 0.0, 1.0]
  end

  def limit
    { effort: 10, velocity: 10, lower: -0.506145, upper: 1.88496 }
  end
end

class JointP2Fingertop
  attr_reader :prefix
  attr_reader :offset

  def initialize(prefix, offset)
    @prefix = prefix
    @offset = offset
  end

  def joint_name
    "#{prefix}_dip"
  end

  def joint_type
    'revolute'
  end

  def origin
    { xyz: [0.0361 + offset, 0.0, -0.0291], rpy: [0.0, 0.0, 1.5708] }
  end

  def parent_link_name
    "#{prefix}_p2_unified"
  end

  def child_link_name
    "#{prefix}_fingertop_unfied"
  end

  def axis
    [0.0, 0.0, 1.0]
  end

  def limit
    { effort: 10, velocity: 10, lower: -0.366519, upper: 2.04204 }
  end
end


class Finger
  attr_reader :prefix
  attr_reader :links
  attr_reader :joints

  def initialize(prefix, links = [], joints = [])
    @prefix = prefix
    @links = links
    @joints = joints
  end
end

def generate_finger(prefix, offset)
  p4 = LinkP4.new(prefix)
  p3 = LinkP3.new(prefix)
  p2 = LinkP2.new(prefix)
  fingertop = LinkFingertop.new(prefix)

  links = [p4, p3, p2, fingertop]
  joints = [
    JointP4P3.new(prefix, offset),
    JointP3P2.new(prefix, offset),
    JointP2Fingertop.new(prefix, offset)
  ]

  Finger.new(prefix, links, joints)
end

def link_urdf(link)
  inertial = link.inertial
  visual = link.visual
  collision = link.collision

  %Q{
  <link name="#{link.link_name}">
    <inertial>
      <origin xyz="#{list_to_string(inertial[:origin][:xyz])}" rpy="#{list_to_string(inertial[:origin][:rpy])}"/>
      <mass value="#{inertial[:mass]}"/>
      <inertia ixx="#{inertial[:inertia][:ixx]}" ixy="#{inertial[:inertia][:ixy]}" ixz="#{inertial[:inertia][:ixz]}" iyy="#{inertial[:inertia][:iyy]}" iyz="#{inertial[:inertia][:iyz]}" izz="#{inertial[:inertia][:izz]}"/>
    </inertial>
    <visual>
      <origin xyz="#{list_to_string(visual[:origin][:xyz])}" rpy="#{list_to_string(visual[:origin][:rpy])}"/>
      <geometry>
        <mesh filename="#{visual[:geometry][:mesh][:filename]}"/>
      </geometry>
      <material name="#{visual[:material][:name]}">
        <color rgba="#{list_to_string(visual[:material][:color][:rgba])}"/>
      </material>
    </visual>
    <collision>
      <origin xyz="#{list_to_string(collision[:origin][:xyz])}" rpy="#{list_to_string(collision[:origin][:rpy])}"/>
      <geometry>
        <mesh filename="#{collision[:geometry][:mesh][:filename]}"/>
      </geometry>
    </collision>
  </link>
  }
end

def joint_urdf(joint)
  limit = joint.limit
  %Q{
  <joint name="#{joint.joint_name}" type="#{joint.joint_type}">
    <origin xyz="#{list_to_string(joint.origin[:xyz])}" rpy="#{list_to_string(joint.origin[:rpy])}"/>
    <parent link="#{joint.parent_link_name}"/>
    <child link="#{joint.child_link_name}"/>
    <axis xyz="#{list_to_string(joint.axis)}"/>
    <limit effort="#{limit[:effort]}" velocity="#{limit[:velocity]}" lower="#{limit[:lower]}" upper="#{limit[:upper]}"/>
  </joint>
  }
end

def render_finger_urdf(finger)
  links_xml = finger.links.map { |l| link_urdf(l) }.join("\n")
  joints_xml = finger.joints.map { |j| joint_urdf(j) }.join("\n")

  %Q{<?xml version="1.0" ?>
<robot name="xela_finger_generated">
#{links_xml}
#{joints_xml}

<link name="base"/>
<joint name="base" type="fixed">
  <origin xyz="0 0 0.2" rpy="1.57 0 0"/>
  <parent link="base"/>
  <child link="#{finger.links[0].link_name}"/>
</joint>
</robot>
}
end

def write_finger_urdf(file_path, finger)
  File.write(file_path, render_finger_urdf(finger))
end


if __FILE__ == $0
  finger = generate_finger('rf', 0.0)
  out_path = ENV['OUT'] || 'finger.urdf'
  write_finger_urdf(out_path, finger)
  puts "Wrote #{out_path}"
end

File: ros_ws/src/leap_description/urdf/finger.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

try:
    # When imported as part of the `xela_description` Python package.
    from .list_to_string import list_to_string
except ImportError:  # pragma: no cover
    # When executed directly as a script.
    from list_to_string import list_to_string


class MCPLink:
    def __init__(self, prefix: str):
        self.prefix = prefix

    # Matching `urdf/robot.urdf` link `mcp_joint`.
    def link_name(self) -> str:
        return f"{self.prefix}_mcp"

    def inertial(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "mass": 0.044,
            "inertia": {
                "ixx": 0.000011499,
                "ixy": 0.000003049,
                "ixz": 0.000000121,
                "iyy": 0.000007796,
                "iyz": 0.000000112,
                "izz": 0.000014759,
            },
        }

    def visual(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.008406902261174496, 0.007766243859716995 , 0.014657354985032912],
                "rpy": [0.0, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/mcp_joint.stl"}},
            "material": {
                "name": "mcp_joint_material",
                "color": {"rgba": [0.866667, 0.866667, 0.0509804, 1.0]},
            },
        }

    def collision(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.008406902261174496, 0.007766243859716995, 0.014657354985032912],
                "rpy": [0.0, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/mcp_joint.stl"}},
        }


class PIPLink:
    def __init__(self, prefix: str):
        self.prefix = prefix

    def link_name(self) -> str:
        return f"{self.prefix}_pip"

    def inertial(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "mass": 0.032,
            "inertia": {
                "ixx": 0.000004798,
                "ixy": 0.000000012,
                "ixz": 0.000000004,
                "iyy": 0.000002862,
                "iyz": 0.000000011,
                "izz": 0.000004234,
            },
        }

    def visual(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.009643363092271328, 0.0003, 0.0007840340104173765],
                "rpy": [-1.5708, -1.5708, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/pip.stl"}},
            "material": {
                "name": "pip_material",
                "color": {"rgba": [0.74902, 0.74902, 0.74902, 1.0]},
            },
        }

    def collision(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.009643363092271328, 0.0003, 0.0007840340104173765],
                "rpy": [-1.5708, -1.5708, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/pip.stl"}},
        }


class DIPLink:
    def __init__(self, prefix: str):
        self.prefix = prefix

    def link_name(self) -> str:
        return f"{self.prefix}_dip"

    def inertial(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "mass": 0.037,
            "inertia": {
                "ixx": 0.000005542,
                "ixy": -0.000000607,
                "ixz": -0.000000017,
                "iyy": 0.000005727,
                "iyz": -0.000000023,
                "izz": 0.000006682,
            },
        }

    def visual(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.021133352895225003, -0.00843211914670488, 0.00978508816209524],
                "rpy": [-3.1416, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/dip.stl"}},
            "material": {
                "name": "dip_material",
                "color": {"rgba": [0.866667, 0.866667, 0.0509804, 1.0]},
            },
        }

    def collision(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.021133352895225003, -0.00843211914670488, 0.00978508816209524],
                "rpy": [-3.1416, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/dip.stl"}},
        }


class FingertipLink:
    def __init__(self, prefix: str):
        self.prefix = prefix

    def link_name(self) -> str:
        return f"{self.prefix}_fingertip"

    def inertial(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "mass": 0.016,
            "inertia": {
                "ixx": 0.000003368,
                "ixy": 0.000000115,
                "ixz": 0.0,
                "iyy": 0.000001556,
                "iyz": 0.0,
                "izz": 0.000002863,
            },
        }

    def visual(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.013286424108533503, -0.006114238386541987, 0.0145],
                "rpy": [3.1416, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/fingertip.stl"}},
            "material": {
                "name": "fingertip_material",
                "color": {"rgba": [0.866667, 0.866667, 0.0509804, 1.0]},
            },
        }

    def collision(self) -> dict[str, Any]:
        return {
            "origin": {
                "xyz": [0.013286424108533503, -0.006114238386541987, 0.0145],
                "rpy": [3.1416, 0.0, 0.0],
            },
            "geometry": {"mesh": {"filename": "package://assets/fingertip.stl"}},
        }


class RELATIPLink:
    def __init__(self, prefix: str):
        self.prefix = prefix

    def link_name(self) -> str:
        return f"{self.prefix}_realtip"

    def inertial(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "mass": 0.1,
            "inertia": {
                "ixx": 0.000025,
                "ixy": 0.0,
                "ixz": 0.0,
                "iyy": 0.000025,
                "iyz": 0.0,
                "izz": 0.000025,
            },
        }

    def visual(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "geometry": {"sphere": {"radius": 0.005}},
            "material": {
                "name": "ball_material",
                "color": {"rgba": [1.0, 0.0, 0.0, 1.0]},
            },
        }

    def collision(self) -> dict[str, Any]:
        return {
            "origin": {"xyz": [0.0, 0.0, 0.0], "rpy": [0.0, 0.0, 0.0]},
            "geometry": {"sphere": {"radius": 0.005}},
        }


class ROTJOINT:
    def __init__(self, prefix: str, offset: float):
        self.prefix = prefix
        self.offset = offset

    def joint_name(self) -> str:
        return f"{self.prefix}_rot"

    def joint_type(self) -> str:
        return "revolute"

    def origin(self) -> dict[str, Any]:
        return {"xyz": [-0.012200000000000007713, 0.038099999999999994982 + self.offset, 0.014500000000000000736], 
        "rpy": [-1.570796326794896558, 0.0, 1.5707963267948943376]}

    def parent_link_name(self) -> str:
        return f"{self.prefix}_mcp"

    def child_link_name(self) -> str:
        return f"{self.prefix}_pip"

    def axis(self) -> list[float]:
        return [0, 0, -1]

    def limit(self) -> dict[str, Any]:
        return {"effort": 0.95, "velocity": 8.48, "lower": -1.047, "upper":1.047}


class PIPJOINT:
    def __init__(self, prefix: str, offset: float):
        self.prefix = prefix
        self.offset = offset

    def joint_name(self) -> str:
        return f"{self.prefix}_pip"

    def joint_type(self) -> str:
        return "revolute"

    def origin(self) -> dict[str, Any]:
        return {"xyz": [0.01499999999999996822, 0.014299999999999904801 + self.offset, -0.012999999999999824196], 
        "rpy": [1.571, -1.571, 0]}

    def parent_link_name(self) -> str:
        return f"{self.prefix}_pip"

    def child_link_name(self) -> str:
        return f"{self.prefix}_dip"

    def axis(self) -> list[float]:
        return [0, 0, -1]

    def limit(self) -> dict[str, Any]:
        return {"effort": 0.95, "velocity": 8.48, "lower": -0.506, "upper": 1.885}


class DIPJOINT:
    def __init__(self, prefix: str, offset: float):
        self.prefix = prefix
        self.offset = offset

    def joint_name(self) -> str:
        return f"{self.prefix}_dip"

    def joint_type(self) -> str:
        return "revolute"

    def origin(self) -> dict[str, Any]:
        return {"xyz": [-4.0880582495572692636e-09, -0.036100004210367367397 + self.offset, 0.00020000000000007858714], 
        "rpy": [0.0, 0.0, 0.0]}

    def parent_link_name(self) -> str:
        return f"{self.prefix}_dip"

    def child_link_name(self) -> str:
        return f"{self.prefix}_fingertip"

    def axis(self) -> list[float]:
        return [0, 0, -1]

    def limit(self) -> dict[str, Any]:
        return {"effort": 0.95, "velocity": 8.48, "lower": -0.366, "upper": 2.042}

class RELATIPJOINT:
    def __init__(self, prefix: str, offset: float):
        self.prefix = prefix
        self.offset = offset

    def joint_name(self) -> str:
        return f"{self.prefix}_realtip"

    def joint_type(self) -> str:
        return "fixed"

    def origin(self) -> dict[str, Any]:
        return {"xyz": [0.02 , -0.07 + self.offset, 0.015], "rpy": [0.0, 0.0, 0.0]}

    def parent_link_name(self) -> str:
        return f"{self.prefix}_fingertip"

    def child_link_name(self) -> str:
        return f"{self.prefix}_realtip"
    def axis(self) -> list[float]:
        return [0, 0, -1]

    def limit(self) -> dict[str, Any]:
        return None
        
    
@dataclass(frozen=True)
class Finger:
    prefix: str
    links: list[Any]
    joints: list[Any]


def generate_finger(prefix: str, offset: float, teleop = True) -> Finger:
    mcp = MCPLink(prefix)
    pip = PIPLink(prefix)
    dip = DIPLink(prefix)
    fingertip = FingertipLink(prefix)
    realtip = RELATIPLink(prefix)

    links = [mcp, pip, dip, fingertip]
    joints = [ ROTJOINT(prefix, offset), PIPJOINT(prefix, offset), DIPJOINT(prefix, offset)]

    if teleop:
        links += [realtip]
        joints += [RELATIPJOINT(prefix, offset)]

    return Finger(prefix, links, joints)


def link_urdf(link: Any) -> str:
    inertial = link.inertial()
    visual = link.visual()
    collision = link.collision()
    
    if "mesh" in visual["geometry"].keys():
        mesh_visual_xml = f"""<mesh filename="{visual["geometry"]["mesh"]["filename"]}"/>"""
    else:
        mesh_visual_xml = f"""<sphere radius="{visual["geometry"]["sphere"]["radius"]}"/>"""
    if "mesh" in collision["geometry"].keys():
        mesh_collision_xml = f"""<mesh filename="{collision["geometry"]["mesh"]["filename"]}"/>"""
    else:
        mesh_collision_xml = f"""<sphere radius="{collision["geometry"]["sphere"]["radius"]}"/>"""

    return f"""
  <link name="{link.link_name()}">
    <inertial>
      <origin xyz="{list_to_string(inertial["origin"]["xyz"])}" rpy="{list_to_string(inertial["origin"]["rpy"])}"/>
      <mass value="{inertial["mass"]}"/>
      <inertia ixx="{inertial["inertia"]["ixx"]}" ixy="{inertial["inertia"]["ixy"]}" ixz="{inertial["inertia"]["ixz"]}" iyy="{inertial["inertia"]["iyy"]}" iyz="{inertial["inertia"]["iyz"]}" izz="{inertial["inertia"]["izz"]}"/>
    </inertial>
    <visual>
      <origin xyz="{list_to_string(visual["origin"]["xyz"])}" rpy="{list_to_string(visual["origin"]["rpy"])}"/>
      <geometry>
        {mesh_visual_xml}
      </geometry>
      <material name="{visual["material"]["name"]}">
        <color rgba="{list_to_string(visual["material"]["color"]["rgba"])}"/>
      </material>
    </visual>
    <collision>
      <origin xyz="{list_to_string(collision["origin"]["xyz"])}" rpy="{list_to_string(collision["origin"]["rpy"])}"/>
      <geometry>
        {mesh_collision_xml}
      </geometry>
    </collision>
  </link>
""".rstrip(
        "\n"
    )


def joint_urdf(joint: Any) -> str:
    limit = joint.limit()
    origin = joint.origin()
    
    limit_xml = ""
    if limit is not None:
        limit_xml = f"""<limit effort="{limit["effort"]}" velocity="{limit["velocity"]}" lower="{limit["lower"]}" upper="{limit["upper"]}"/>"""

     
    return f"""
  <joint name="{joint.joint_name()}" type="{joint.joint_type()}">
    <origin xyz="{list_to_string(origin["xyz"])}" rpy="{list_to_string(origin["rpy"])}"/>
    <parent link="{joint.parent_link_name()}"/>
    <child link="{joint.child_link_name()}"/>
    <axis xyz="{list_to_string(joint.axis())}"/>
    {limit_xml}
  </joint>
""".rstrip(
        "\n"
    )


def render_finger_urdf(finger: Finger) -> str:
    links_xml = "\n".join(link_urdf(l) for l in finger.links)
    joints_xml = "\n".join(joint_urdf(j) for j in finger.joints)

    return f"""<?xml version="1.0" ?>
<robot name="xela_finger_generated">
{links_xml}
{joints_xml}

<link name="base"/>
<joint name="base" type="fixed">
  <origin xyz="0 0 0.2" rpy="1.57 0 0"/>
  <parent link="base"/>
  <child link="{finger.links[0].link_name()}"/>
</joint>
</robot>
"""


def write_finger_urdf(file_path: str, finger: Finger) -> None:
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(render_finger_urdf(finger))


if __name__ == "__main__":
    import os

    finger = generate_finger("rf", 0.0)
    out_path = os.environ.get("OUT") or "finger.urdf"
    write_finger_urdf(out_path, finger)
    print(f"Wrote {out_path}")

